// gpt-turbo/dist/classes/ConversationPluginService.js
/**
 * Manages the plugins of a conversation and acts as a bridge between the plugins and the conversation.
 *
 * @remarks
 * Most of these methods are not documented (with TypeDocs).
 * In most cases, this is because they just call the equivalent method on each plugin.
 * Refer to the documentation of the `ConversationPluginCreatorDefinitionBase` interface for their documentation.
 *
 * @internal
 * This class is used internally by the library and is not meant to be **instantiated** by consumers of the library.
 */
export class ConversationPluginService {
    pluginCreators;
    plugins = [];
    constructor(pluginCreators) {
        this.pluginCreators = pluginCreators;
    }
    getPlugins() {
        return this.plugins.slice();
    }
    onInit(properties, pluginsData) {
        for (const pluginCreator of this.pluginCreators) {
            if (this.plugins.some((p) => p.name === pluginCreator.name)) {
                console.warn(`[gpt-turbo] Skipping duplicate plugin "${pluginCreator.name}"`);
                continue;
            }
            this.plugins.push({
                ...pluginCreator.creator(properties, pluginsData?.[pluginCreator.name]),
                name: pluginCreator.name,
            });
        }
        this.onPostInit();
    }
    onPostInit() {
        this.plugins.forEach((p) => p.onPostInit?.());
    }
    transformConversationJson(json) {
        return this.plugins.reduce((json, p) => p.transformConversationJson?.(json) ?? json, json);
    }
    transformConversationCallableFunctionsJson(json) {
        return this.plugins.reduce((json, p) => p.transformConversationCallableFunctionsJson?.(json) ?? json, json);
    }
    transformConversationConfigJson(json) {
        return this.plugins.reduce((json, p) => p.transformConversationConfigJson?.(json) ?? json, json);
    }
    transformConversationHistoryJson(json) {
        return this.plugins.reduce((json, p) => p.transformConversationHistoryJson?.(json) ?? json, json);
    }
    transformConversationRequestOptionsJson(json) {
        return this.plugins.reduce((json, p) => p.transformConversationRequestOptionsJson?.(json) ?? json, json);
    }
    async onUserPrompt(message) {
        for (const plugin of this.plugins) {
            await plugin.onUserPrompt?.(message);
        }
    }
    async onUserPromptError(error) {
        for (const plugin of this.plugins) {
            try {
                await plugin.onUserPromptError?.(error);
            }
            catch (e) {
                console.error(`[gpt-turbo] Plugin "${plugin.name}" errored during "onUserPromptError":`);
                console.error(e);
            }
        }
    }
    async onChatCompletion(message) {
        for (const plugin of this.plugins) {
            await plugin.onChatCompletion?.(message);
        }
    }
    async transformFunctionResult(result) {
        return this.plugins.reduce(async (result, p) => p.transformFunctionResult?.(await result) ?? result, Promise.resolve(result));
    }
    async onFunctionPrompt(message) {
        for (const plugin of this.plugins) {
            await plugin.onFunctionPrompt?.(message);
        }
    }
    async onFunctionPromptError(error) {
        for (const plugin of this.plugins) {
            try {
                await plugin.onFunctionPromptError?.(error);
            }
            catch (e) {
                console.error(`[gpt-turbo] Plugin "${plugin.name}" errored during "onFunctionPromptError":`);
                console.error(e);
            }
        }
    }
    getPluginsData() {
        return this.plugins.reduce((data, p) => {
            if (p.getPluginData !== undefined) {
                data[p.name] = p.getPluginData();
            }
            return data;
        }, {});
    }
    async onModeration(message) {
        for (const plugin of this.plugins) {
            await plugin.onModeration?.(message);
        }
    }
}
