// gpt-turbo/dist/classes/ConversationRequestOptions.js
import { conversationRequestOptionsSchema, } from "../schemas/conversationRequestOptions.schema.js";
/**
 * Holds the callable functions of a `Conversation`.
 *
 * @internal
 * This class is used internally by the library and is not meant to be **instantiated** by consumers of the library.
 */
export class ConversationRequestOptions {
    pluginService;
    headers;
    proxy;
    constructor(pluginService, options = {}) {
        this.pluginService = pluginService;
        this.setRequestOptions(options);
    }
    /**
     * Serializes the `ConversationRequestOptions` to JSON.
     *
     * @returns A JSON representation of the `ConversationRequestOptions` instance.
     */
    toJSON() {
        const json = this.getRequestOptions();
        return conversationRequestOptionsSchema.parse(this.pluginService.transformConversationRequestOptionsJson(json));
    }
    /**
     * Returns the current request options.
     */
    getRequestOptions() {
        return {
            headers: this.headers,
            proxy: this.proxy,
        };
    }
    /**
     * Sets new request options to be used as defaults for all HTTP requests made by this conversation.
     *
     * @param requestOptions The new request options to use.
     * @param merge Set to `true` to shallow merge the new request options with the existing request options instead of replacing them.
     */
    setRequestOptions(requestOptions, merge = false) {
        const options = merge
            ? { ...this.headers, ...this.proxy, ...requestOptions }
            : requestOptions;
        this.headers = options.headers;
        this.proxy = options.proxy;
    }
}
