// gpt-turbo/dist/classes/Message.js
import { v4 as uuid } from "uuid";
import { messageSchema } from "../schemas/message.schema.js";
import createModeration from "../utils/createModeration.js";
import { EventManager } from "./EventManager.js";
/**
 * A message in a Conversation.
 */
export class Message {
    /**
     * A UUID generated for this message by the library.
     */
    id = uuid();
    _role;
    _model;
    _content;
    _name;
    _functionCall;
    _flags = null;
    _isStreaming = false;
    messageUpdateEvents = new EventManager();
    messageStreamingEvents = new EventManager();
    messageStreamingStartEvents = new EventManager();
    messageStreamingStopEvents = new EventManager();
    messageContentStreamEvents = new EventManager();
    /**
     * Creates a new Message instance.
     *
     * @param role The role of who this message is from. Either "user", "assistant" or "system".
     * @param content The content of the message.
     * @param model The model used for processing this message.
     */
    constructor(role = "user", content = "", model = "") {
        this.role = role;
        this.model = model;
        this.content = content;
    }
    /**
     * Creates a new `Message` instance from a serialized message.
     *
     * @param json The JSON object of the Message instance.
     * @returns A new `Message` instance
     */
    static fromJSON(json) {
        const messageJson = messageSchema.parse(json);
        const message = new Message(messageJson.role, messageJson.content, messageJson.model);
        if (messageJson.id)
            message.id = messageJson.id;
        if (messageJson.flags)
            message.flags = messageJson.flags;
        return message;
    }
    /**
     * Serializes the message to JSON.
     *
     * @returns The `Message` as a JSON object.
     */
    toJSON() {
        const json = {
            id: this.id,
            role: this.role,
            content: this.content,
            model: this.model,
            name: this._name,
            function_call: this._functionCall,
            flags: this.flags,
        };
        return messageSchema.parse(json);
    }
    /**
     * Call the OpenAI moderation API to check if the message is flagged. Only called once for the same content.
     *
     * @param apiKey The OpenAI API key
     * @param requestOptions The request options to pass to fetch
     * @returns The flags applied on the message
     */
    async moderate(apiKey, requestOptions = {}) {
        const flags = this.flags;
        if (flags) {
            return flags;
        }
        if (!this.content) {
            this.flags = [];
            return this.flags;
        }
        const response = await createModeration({
            apiKey,
            input: this.content,
        }, requestOptions);
        const { flagged, categories } = response.results[0];
        this.flags = flagged
            ? Object.keys(categories).filter((category) => categories[category])
            : [];
        return this.flags;
    }
    /**
     * Progessively adds content to the message from a streamed response type.
     *
     * @param response The ReadableStream from the OpenAI API
     */
    async readContentFromStream(response) {
        try {
            this.isStreaming = true;
            const reader = response.getReader();
            const decoder = new TextDecoder();
            let functionCallName = "";
            let functionCallArguments = "";
            while (true) {
                const { done, value } = await reader.read();
                if (done)
                    break;
                const decoded = decoder.decode(value).trim();
                const chunks = decoded.split("data: ").filter((c) => c.trim());
                for (const chunk of chunks) {
                    try {
                        const json = JSON.parse(chunk);
                        const content = json?.choices?.[0]?.delta?.content ?? null;
                        const functionCall = json?.choices?.[0]?.delta?.function_call ?? null;
                        if (functionCall) {
                            const { name, arguments: argsStr } = functionCall;
                            functionCallName += name ?? "";
                            functionCallArguments += argsStr ?? "";
                            let args = this.functionCall
                                ? { ...this.functionCall.arguments }
                                : {};
                            try {
                                args = JSON.parse(functionCallArguments);
                            }
                            finally {
                                this.functionCall = {
                                    name: functionCallName,
                                    arguments: args,
                                };
                            }
                        }
                        else {
                            if (!content)
                                continue;
                            this.content += content;
                        }
                    }
                    catch {
                        continue;
                    }
                }
            }
        }
        catch (err) {
            console.error(err);
        }
        finally {
            this.isStreaming = false;
        }
    }
    /**
     * Whether the message is a function call by the assistant
     */
    isFunctionCall() {
        return (this.role === "assistant" &&
            this.content === null &&
            this.functionCall !== undefined);
    }
    /**
     * Whether the message is a function call result by the user
     */
    isFunction() {
        return this.role === "function" && this.name !== undefined;
    }
    /**
     * Whether the message is a regular chat completion message
     */
    isCompletion() {
        return (this.role !== "function" &&
            this.content !== null &&
            this.functionCall === undefined &&
            this.name === undefined);
    }
    /** The role of who this message is from. */
    get role() {
        return this._role;
    }
    set role(role) {
        this._role = role;
    }
    /** The model used to generate this message. */
    get model() {
        return this._model;
    }
    set model(model) {
        this._model = model;
    }
    /** The content of the message. */
    get content() {
        return this._content;
    }
    set content(content) {
        this._content = content;
        this.flags = null;
        this.messageUpdateEvents.emit(content, this);
        if (this.isStreaming) {
            this.messageContentStreamEvents.emit(content, this.isStreaming, this);
        }
    }
    get name() {
        return this._name;
    }
    set name(name) {
        this._name = name;
    }
    get functionCall() {
        return this._functionCall;
    }
    set functionCall(functionCall) {
        this._functionCall = functionCall;
        this.flags = null;
        this.content = null; // also call notifyMessageUpdate() to notify listeners
    }
    /** The flags detected by OpenAI's moderation API. Only set after calling `moderate`. */
    get flags() {
        return this._flags?.slice() ?? null;
    }
    set flags(flags) {
        this._flags = flags;
    }
    /** Whether the message is flagged by OpenAI's moderation API. Always `false` unless `moderate` has been called. */
    get isFlagged() {
        return (this.flags?.length ?? 0) > 0;
    }
    /** Whether the message is currently being streamed. */
    get isStreaming() {
        return this._isStreaming;
    }
    set isStreaming(isStreaming) {
        if (this._isStreaming === isStreaming)
            return;
        this._isStreaming = isStreaming;
        this.messageStreamingEvents.emit(isStreaming, this);
        isStreaming
            ? this.messageStreamingStartEvents.emit(this)
            : this.messageStreamingStopEvents.emit(this);
        this.messageContentStreamEvents.emit(this.content, isStreaming, this);
        if (!isStreaming) {
            this.messageContentStreamEvents.clear();
        }
    }
    /**
     * JSON definition of the message as consumed by OpenAI's API.
     */
    get chatCompletionMessage() {
        if (this.isFunctionCall()) {
            return {
                role: this.role,
                content: this.content,
                function_call: {
                    name: this.functionCall.name,
                    arguments: JSON.stringify(this.functionCall.arguments),
                },
            };
        }
        if (this.isFunction()) {
            return {
                content: this.content,
                name: this.name,
                role: this.role,
            };
        }
        if (this.isCompletion()) {
            return {
                content: this.content,
                role: this.role,
            };
        }
        throw new Error("Message type not recognized.");
    }
    /**
     * Add a listener for message content changes.
     *
     * @param listener The listener to trigger when `content` changes
     * @returns An unsubscribe function for this `listener`
     */
    onUpdate(listener) {
        return this.messageUpdateEvents.addListener(listener);
    }
    /**
     * Add a listener that is called only once when the message content changes.
     *
     * @param listener The listener to trigger when `content` changes
     * @returns An unsubscribe function for this `listener`
     */
    onceUpdate(listener) {
        return this.messageUpdateEvents.once(listener);
    }
    /**
     * Removes a message update listener, previously set with `onUpdate`.
     *
     * @param listener The previously added listener
     */
    offUpdate(listener) {
        return this.messageUpdateEvents.removeListener(listener);
    }
    /**
     * Adds a listener for message streaming state changes.
     *
     * @param listener The listener to trigger when `isStreaming` changes
     * @returns An unsubscribe function for this `listener`
     */
    onStreamingUpdate(listener) {
        return this.messageStreamingEvents.addListener(listener);
    }
    /**
     * Adds a listener that is called only once when the message streaming state changes.
     *
     * @param listener The listener to trigger when `isStreaming` changes
     * @returns An unsubscribe function for this `listener`
     */
    onceStreamingUpdate(listener) {
        return this.messageStreamingEvents.once(listener);
    }
    /**
     * Removes a message streaming listener, previously set with `onStreamingUpdate`.
     *
     * @param listener The previously added listener
     */
    offStreaming(listener) {
        return this.messageStreamingEvents.removeListener(listener);
    }
    /**
     * Adds a listener for message streaming start.
     *
     * @param listener The listener to trigger when `isStreaming` is set to `true`
     * @returns An unsubscribe function for this `listener`
     */
    onStreamingStart(listener) {
        return this.messageStreamingStartEvents.addListener(listener);
    }
    /**
     * Adds a listener that is called only once when the message streaming starts.
     *
     * @param listener The listener to trigger when `isStreaming` is set to `true`
     * @returns An unsubscribe function for this `listener`
     */
    onceStreamingStart(listener) {
        return this.messageStreamingStartEvents.once(listener);
    }
    /**
     * Removes a message streaming start listener, previously set with `onStreamingStart`.
     *
     * @param listener The previously added listener
     */
    offStreamingStart(listener) {
        return this.messageStreamingStartEvents.removeListener(listener);
    }
    /**
     * Adds a listener for message streaming stop.
     *
     * @param listener The listener to trigger when `isStreaming` is set to `false`
     * @returns An unsubscribe function for this `listener`
     */
    onStreamingStop(listener) {
        return this.messageStreamingStopEvents.addListener(listener);
    }
    /**
     * Adds a listener that is called only once when the message streaming stops.
     *
     * @param listener The listener to trigger when `isStreaming` is set to `false`
     * @returns An unsubscribe function for this `listener`
     */
    onceStreamingStop(listener) {
        return this.messageStreamingStopEvents.once(listener);
    }
    /**
     * Removes a message streaming stop listener, previously set with `onStreamingStop`.
     *
     * @param listener The previously added listener
     */
    offStreamingStop(listener) {
        return this.messageStreamingStopEvents.removeListener(listener);
    }
    /**
     * Adds a listener that is fired whenever the message content is updated during streaming. Also fires when streaming starts/ends.
     *
     * @remarks
     * Unlike the other listeners which behave like normal event listeners, this special listener is unsubscribed automatically when streaming ends.
     * If this is not desired, use `onUpdate` and `onStreamingStart`/`onStreamingStop` instead.
     *
     * @param listener The listener to trigger when `content` changes during streaming
     * @returns An unsubscribe function for this `listener`
     */
    onContentStream(listener) {
        return this.messageContentStreamEvents.addListener(listener);
    }
    /**
     * Adds a listener that is fired only once whenever the message content is updated during streaming. Also fires when streaming starts/ends.
     *
     * @remarks
     * Unlike the other listeners which behave like normal event listeners, this special listener is unsubscribed automatically when streaming ends, regardless of whether it was called once or not.
     * If this is not desired, use `onceUpdate` and `onceStreamingStart`/`onceStreamingStop` instead.
     *
     * @param listener The listener to trigger when `content` changes during streaming
     * @returns An unsubscribe function for this `listener`
     */
    onceContentStream(listener) {
        return this.messageContentStreamEvents.once(listener);
    }
    /**
     * Removes a message streaming content listener, previously set with `onContentStream`.
     *
     * @remarks
     * You do not need to call this method manually, as this listener automatically unsubscribes all listeners when streaming ends.
     * You should use this if you want to unsubscribe a listener before streaming ends.
     *
     * @param listener The previously added listener
     */
    offContentStream(listener) {
        return this.messageContentStreamEvents.removeListener(listener);
    }
}
