// gpt-turbo/dist/classes/CallableFunctionParameters/CallableFunctionParameter.js
import { v4 as uuid } from "uuid";
import { jsonSchemaBaseSchema, } from "../../schemas/jsonSchema.schema.js";
export class CallableFunctionParameter {
    name;
    /**
     * A UUID generated for this parameter by the library.
     */
    id = uuid();
    title;
    description;
    default;
    examples;
    readOnly;
    writeOnly;
    deprecated;
    $comment;
    constructor(name, options = {}) {
        this.name = name;
        this.name = name;
        this.title = options.title;
        this.description = options.description;
        this.default = options.default;
        this.examples = options.examples;
        this.readOnly = options.readOnly;
        this.writeOnly = options.writeOnly;
        this.deprecated = options.deprecated;
        this.$comment = options.$comment;
    }
    toJSONBase() {
        const json = {
            title: this.title,
            description: this.description,
            default: this.default,
            examples: this.examples,
            readOnly: this.readOnly,
            writeOnly: this.writeOnly,
            deprecated: this.deprecated,
            $comment: this.$comment,
        };
        return jsonSchemaBaseSchema.parse(json);
    }
}
