// gpt-turbo/dist/classes/CallableFunctionParameters/CallableFunctionArray.js
import { jsonSchemaArraySchema, } from "../../schemas/jsonSchema.schema.js";
import { CallableFunctionParameter } from "./CallableFunctionParameter.js";
import { CallableFunctionParameterFactory } from "./CallableFunctionParameterFactory.js";
export class CallableFunctionArray extends CallableFunctionParameter {
    type = "array";
    minContains;
    maxContains;
    minItems;
    maxItems;
    uniqueItems;
    items;
    prefixItems;
    contains;
    constructor(name, options = {}) {
        super(name, options);
        this.minContains = options.minContains;
        this.maxContains = options.maxContains;
        this.minItems = options.minItems;
        this.maxItems = options.maxItems;
        this.uniqueItems = options.uniqueItems;
    }
    static fromJSON(name, json) {
        const callableFunctionArrayModel = jsonSchemaArraySchema.parse(json);
        const callableFunctionArray = new CallableFunctionArray(name, callableFunctionArrayModel);
        const { items, prefixItems, contains } = callableFunctionArrayModel;
        if (items) {
            if (typeof items === "boolean") {
                callableFunctionArray.items = items;
            }
            else {
                callableFunctionArray.items =
                    CallableFunctionParameterFactory.fromJSON(`${name}.items`, items);
            }
        }
        if (prefixItems) {
            callableFunctionArray.prefixItems = [];
            for (const value of prefixItems) {
                callableFunctionArray.prefixItems.push(CallableFunctionParameterFactory.fromJSON(`${name}.prefixItems`, value));
            }
        }
        if (contains) {
            callableFunctionArray.contains =
                CallableFunctionParameterFactory.fromJSON(`${name}.contains`, contains);
        }
        return callableFunctionArray;
    }
    /**
     * Whether the given `JsonSchema` is a `JsonSchemaArray`.
     *
     * Note: This method does not check if the schema is valid, only if it has properties unique to array schemas.
     */
    static isArraySchema(json) {
        return json.type === "array";
    }
    toJSON() {
        const json = {
            ...this.toJSONBase(),
            type: this.type,
            minContains: this.minContains,
            maxContains: this.maxContains,
            minItems: this.minItems,
            maxItems: this.maxItems,
            uniqueItems: this.uniqueItems,
        };
        if (this.items) {
            if (typeof this.items === "boolean") {
                json.items = this.items;
            }
            else {
                json.items = this.items.toJSON();
            }
        }
        if (this.prefixItems) {
            json.prefixItems = [];
            for (const value of this.prefixItems) {
                json.prefixItems.push(value.toJSON());
            }
        }
        if (this.contains) {
            json.contains = this.contains.toJSON();
        }
        return jsonSchemaArraySchema.parse(json);
    }
}
