// gpt-turbo/dist/classes/CallableFunctionParameters/CallableFunctionString.js
import { jsonSchemaStringSchema, } from "../../schemas/jsonSchema.schema.js";
import { CallableFunctionParameter } from "./CallableFunctionParameter.js";
export class CallableFunctionString extends CallableFunctionParameter {
    type = "string";
    minLength;
    maxLength;
    pattern;
    enum;
    format;
    constructor(name, options = {}) {
        super(name, options);
        this.minLength = options.minLength;
        this.maxLength = options.maxLength;
        this.pattern = options.pattern;
        this.enum = options.enum;
        this.format = options.format;
    }
    static fromJSON(name, json) {
        const callableFunctionStringModel = jsonSchemaStringSchema.parse(json);
        const callableFunctionString = new CallableFunctionString(name, callableFunctionStringModel);
        return callableFunctionString;
    }
    /**
     * Whether the given `JsonSchema` is a `JsonSchemaString`.
     *
     * Note: This method does not check if the schema is valid, only if it has properties unique to string schemas.
     */
    static isStringSchema(json) {
        return json.type === "string";
    }
    toJSON() {
        const json = {
            ...this.toJSONBase(),
            type: this.type,
            minLength: this.minLength,
            maxLength: this.maxLength,
            pattern: this.pattern,
            enum: this.enum,
            format: this.format,
        };
        return jsonSchemaStringSchema.parse(json);
    }
}
