// gpt-turbo/dist/classes/CallableFunctionParameters/CallableFunctionNumber.js
import { jsonSchemaNumberSchema, } from "../../schemas/jsonSchema.schema.js";
import { CallableFunctionParameter } from "./CallableFunctionParameter.js";
export class CallableFunctionNumber extends CallableFunctionParameter {
    type = "number";
    multipleOf;
    minimum;
    exclusiveMinimum;
    maximum;
    exclusiveMaximum;
    constructor(name, options = {}) {
        super(name, options);
        this.multipleOf = options.multipleOf;
        this.minimum = options.minimum;
        this.exclusiveMinimum = options.exclusiveMinimum;
        this.maximum = options.maximum;
        this.exclusiveMaximum = options.exclusiveMaximum;
    }
    static fromJSON(name, json) {
        const callableFunctionNumberModel = jsonSchemaNumberSchema.parse(json);
        const callableFunctionNumber = new CallableFunctionNumber(name, callableFunctionNumberModel);
        return callableFunctionNumber;
    }
    /**
     * Whether the given `JsonSchema` is a `JsonSchemaNumber`.
     *
     * Note: This method does not check if the schema is valid, only if it has properties unique to number schemas.
     */
    static isNumberSchema(json) {
        return json.type === "number";
    }
    toJSON() {
        const json = {
            ...this.toJSONBase(),
            type: this.type,
            multipleOf: this.multipleOf,
            minimum: this.minimum,
            exclusiveMinimum: this.exclusiveMinimum,
            maximum: this.maximum,
            exclusiveMaximum: this.exclusiveMaximum,
        };
        return jsonSchemaNumberSchema.parse(json);
    }
}
